### PR TITLE
Expanded on testing for JSON processing and tuple class hierarchy flattened

### PR DESCRIPTION
**Pull request by mrw8**
Expanded testing for JSON parsing and writing to have a different test for each tuple type. Docstrings were added to the remaining tuple classes to say what each tuple represents and the fields inside of the tuple. 
Lastly, the class hierarchy was flattened (besides the RtTuple super class), as the previous class hierarchy if extended to its logical conclusion would require a large amount of multiple inheritance that is less maintainable than a simply flattened version.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -22,8 +22,6 @@ requires-python = ">=3.11"
 
 [tool.pytest.ini_options]
 pythonpath = [
```
Comment (hoganwr) [inline pyproject.toml:24]: I see, this is why pytest worked and python3 tests/test_atuple.py does not.

File: src/rt_core_v2/rttuple.py [hunk 2]
```diff
@@ -25,53 +26,55 @@ class TupleType(ValueEnum):
...
 class PorType(ValueEnum):
 	singular = '+SU'
 	non_singular = '-SU'
 
+#TODO Move this into the classes, as it is not a semantically sound placement here
 class TupleComponents(enum.Enum):
+	ruit = 'ruit'
+	type = 'type'
 	ar = 'ar'
-	C = 'C'
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple.py:45]: Most of these deletions and additions appear to just be re-ordering, so it's hard to pick out what's new and what if anything is really going out.

File: src/rt_core_v2/rttuple.py [hunk 3]
```diff
@@ -25,53 +26,55 @@ class TupleType(ValueEnum):
...
 class RtTuple(ABC):
+	"""Abstract Referent Tracking tuple that contains the information that all referent tracking tuples contain
+	
+	Attributes:
+	ruit -- The rui of this tuple
+	type -- The id of the tuple component
+	"""
 	tuple_type = None
-	params = {TupleComponents.ruit:TupleComponents.ruit.value, TupleComponents.type:TupleComponents.type.value}
+	params = {**enum_to_dict({TupleComponents.ruit, TupleComponents.type})}
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple.py:77]: I'm not familiar with the double asterisk notation. I'll have to go look it up. There's no documentation explaining this change in design.

File: src/rt_core_v2/rttuple.py [hunk 4]
```diff
@@ -90,21 +93,22 @@ def get_str_attributes(self):
...
 	"""Referent Tracking assignment tuple that registers assignment of an RUI to a PoR
 	
 	Attributes:
 	ar -- The status of ruip
 	ruip -- The Rui that is being assigned for the first time
-	ruia -- The Rui of the author of this Atuple
-	unique -- 
-	t -- The time of the creation of the Atuple
+	ruia -- The Rui of the author of this ATuple
+	unique -- Asserts whether this is a non-repeatable or repeatable portion of reality
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple.py:103]: That is not correct. It asserts whether the person assigning the RUI believes that there is an exact one-to-one correspondence between the RUI and the entity (+SU or "singularly unique) or whether it could be that some other RUI also denotes the entity (-SU or "not singularly unique).

File: src/rt_core_v2/rttuple.py [hunk 5]
```diff
@@ -90,21 +93,22 @@ def get_str_attributes(self):
...
+	t -- The time of the creation of the ATuple
 	"""
-	params = {**RtTuple.params, **enum_to_dict({TupleComponents.ar, TupleComponents.t, TupleComponents.ruia, TupleComponents.unique, TupleComponents.ruip})}
+	params = {**RtTuple.params, **enum_to_dict({TupleComponents.ar, TupleComponents.t,
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple.py:106]: I don't understand specifying all the member variables of a tuple using this mechanism vs. just as member variables in the class. Perhaps this is a typical Pythonic way of doing things, in which case I'd like to learn more before approving.

File: src/rt_core_v2/rttuple.py [hunk 6]
```diff
@@ -152,224 +156,291 @@ def create_assigned(self):
...
 
-# This class is the superclass of all Nto* tuples. They all relate some
-#	non-repeatable portion of reality to some portion of reality (in 
-#	some cases a repeatable PoR, and in others non-repeatable ones) or
-#	in the case of NtoC, it is asserting that the N is "annotated by"
-#	the "concept" from some concept system.
-# We require the NtoXGenericTuple to accomodate NtoLackR
-#	All the other Nto* tuples extend NtoXTuple
-#
-class NtoXGenericTuple(RtTuple):
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple.py:168]: This bit is hard to follow. The way the diff gets displayed it looks like you're changing the name of a class (from NtoXGenericTuple to DTuple), but somehow you've moved the delaration of the DTuple class up above where NtoXTuple used to be but left its guts below.  Weird.

File: src/rt_core_v2/rttuple.py [hunk 7]
```diff
@@ -152,224 +156,291 @@ def create_assigned(self):
...
-		return not self.polarity
+	tuple_type = TupleType.F
+	params = {**RtTuple.params, **enum_to_dict({TupleComponents.ruid, TupleComponents.ruia, TupleComponents.ta, TupleComponents.C})}
 	
+	def __init__(self, ruid: Rui=None, ruit: Rui=None, ta: TempRef=None, ruia: Rui=None, C: float=1.0):
+		super().__init__(ruid)
+		self.ruit_ref = ruit
+		self.ruia = ruia
+		self.ta = ta
+		self.C = C
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple.py:227]: Here's the capital letter 'C' appearing in the same font color as various class names.  Why is it a float??

File: src/rt_core_v2/rttuple.py [hunk 8]
```diff
@@ -152,224 +156,291 @@ def create_assigned(self):
...
 		attributes[self.params[TupleComponents.ruidt]] = str(self.ruidt)
 		return attributes
 
-class NtoLackR(NtoXGenericTuple):
-	"""Tuple type that asserts that for all instances of a given type, a specific
-		non-repeatable portion of reality is not related to any of them by a 
-		given relation"""
-	#NtoR(-) -tuple NtoR(-)#< r, RUIp, RUIr, rT/‘-’, tr/‘-’ >
+class NtoLackRTuple(RtTuple):
+	"""Tuple type that relates a non-repeatable portion of reality to a repeatable portion of reality in a negative relationship
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple.py:402]: This change in description doesn't seem accurate to me. Is this text taken from Werner's writing?

File: src/rt_core_v2/rttuple.py [hunk 9]
```diff
@@ -152,224 +156,291 @@ def create_assigned(self):
...
-		if not self.params:
-			self.params[TupleComponents.ruir] = TupleComponents.ruir.value
-			self.params[TupleComponents.tr] = TupleComponents.tr.value
+	def __init__(self, ruit: Rui, r: str, ruip: Rui, ruir: Rui, rT, tr: TempRef):
+		super().__init__(ruit)
+		self.relation = r
+		self.ruip = ruip
 		self.ruir = ruir
-		self.tr = tr 
+		self.time_relation = rT
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple.py:423]: It's not a "time relation" given how we've implemented it. I'm not following Werner's scheme for these parameters. We really do want to require a temporal region here ultimately.

File: src/rt_core_v2/rttuple.py [hunk 10]
```diff
@@ -152,224 +156,291 @@ def create_assigned(self):
...
-
-		return attributes
\ No newline at end of file
+"""Mapping from tuple id to the corresponding tuple class"""
+type_to_class = {
+	TupleType.A: ATuple,
+	TupleType.D: DTuple,
+	TupleType.F: FTuple,
+	TupleType.NtoDE: NtoDETuple,
+	TupleType.NtoNTuple: NtoNTuple,
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple.py:442]: Why add "Tuple" here but not on the others?

File: tests/test_code.py [hunk 11]
```diff
@@ -1,4 +1,4 @@
-from ids_codes import Concept, Rui
+from rt_core_v2.ids_codes import Concept, Rui
 import sys
```
Comment (hoganwr) [inline tests/test_code.py:2]: LOL I thought I got rid of all of the import sys, sys.path.append problems. They shouldn't be needed if our project is configured correctly.

File: src/rt_core_v2/rttuple_formatter.py [hunk 12]
```diff
@@ -1,13 +1,22 @@
...
+import json
+import enum
+from io import StringIO
+
+from rt_core_v2.rttuple import TupleComponents, TupleType, type_to_class
 
 def rttuple_to_json(input_rttuples):
     """Converts either a list or an RtTuple to a json object"""
-    return json.dumps(input_rttuples.get_str_attributes())
+    return json.dumps(input_rttuples.get_str_attributes(), default=lambda o: o.toJSON() if hasattr(o, 'toJSON') else str(o))
```
Comment (hoganwr) [inline src/rt_core_v2/rttuple_formatter.py:9]: Why not:

for rttuple in input_rttuples:
    json.dumps(rttuple.__dict__)

Then perhaps all the parameters() things going on are not as necessary? 

And that could elminate dependency of rttuple.py on json

Comment (hoganwr) [reply src/rt_core_v2/rttuple_formatter.py:9]: Of course there are potential design reasons to leave it up to each tuple class on how its instances convert themselves to JSON.

Comment (hoganwr) [reply src/rt_core_v2/rttuple_formatter.py:9]: But I don't think so, we just want them to write out all their parameters, and AFAIK, we don't have any member variables that aren't tuple parameters (i.e., they're not part of RT2 spec but there for convenience sake in processing etc.)

Comment (hoganwr) [reply src/rt_core_v2/rttuple_formatter.py:9]: Another possible consideration is if we want to convert datatypes in writing them out to JSON.  For example, if we had a boolean member variable that we wanted to write out not as Bool in JSON, but as string.

Comment (hoganwr) [reply src/rt_core_v2/rttuple_formatter.py:9]: For example the positive / negative (+/-) parameter in all NtoX tuples (except NtoR-)

File: src/rt_core_v2/rttuple.py (hunk 7 above)
Comment (hoganwr) [reply src/rt_core_v2/rttuple.py:227]: nevermind I see why it's a float

File: src/rt_core_v2/rttuple.py (hunk 3 above)
Comment (hoganwr) [reply src/rt_core_v2/rttuple.py:77]: Looking further, it looks like this is how you're reconstituting tuples from JSON.

File: src/rt_core_v2/rttuple.py (hunk 5 above)
Comment (hoganwr) [reply src/rt_core_v2/rttuple.py:106]: Again, looks like the way to rebuild them from JSON

File: tests/test_ntoxtuple.py [hunk 13]
```diff
@@ -39,27 +38,27 @@ def print_nton_tuple(nton):
 h = Rui.Rui()
 # Rui that stands for interval over which author has been instance of human being
 tr1 = Rui.Rui()
-k = Atuple(tr1, ruia=a)
+k = ATuple(tr1, ruia=a)
 
-ntor = NtoR(None, a, True, "instance of", h, tr1)
-print_ntor_tuple(ntor)
+NtoRTuple = NtoRTuple(None, True, "part of", a, h, "Part of", Rui.TempRef())
```
Comment (hoganwr) [inline tests/test_ntoxtuple.py:43]: Why is "part of" here twice?

File: tests/test_ntoxtuple.py [hunk 14]
```diff
@@ -39,27 +38,27 @@ def print_nton_tuple(nton):
...
+w = ATuple(s, ruia=a, unique="+SU")
+print_ATuple(w)
 
 #let tr2 be interval over which kuala lumpur has been part of Malaysia
 tr2 = Rui.Rui()
-j = Atuple(tr2, ruia=a)
-nton = NtoN(None, x, True, "part of", [x, s], tr2)
-print_nton_tuple(nton)
+j = ATuple(tr2, ruia=a)
+NtoNTuple = NtoNTuple(None, True, "part of", [x, s], "part of", Rui.TempRef())
```
Comment (hoganwr) [inline tests/test_ntoxtuple.py:63]: also here: why is "part of" here twice?

**Review by hoganwr — APPROVED** [review]

File: src/rt_core_v2/rttuple.py (hunk 2 above)
Comment (mrw8) [reply src/rt_core_v2/rttuple.py:45]: These changes were just a reordering, so that entries in the enum were closer to other semantically related entries.

File: src/rt_core_v2/rttuple.py (hunk 3 above)
Comment (mrw8) [reply src/rt_core_v2/rttuple.py:77]: This is how I'm enumerating the tuple entries for each class. The ** allows for unpacking a dictionary so that the entries can be added to another. For that instance, it is not necessary, as there is not another dictionary that is being merged with it. It is done purely to be consistent in format with the other tuples, but I'd be happy to change it.

I felt that calling the helper function looks cleaner than having each entry being in the form:
 "TupleComponents.x : TupleComponents.x.value

File: src/rt_core_v2/rttuple.py (hunk 4 above)
Comment (mrw8) [reply src/rt_core_v2/rttuple.py:103]: That makes more sense. I was a bit unclear as to what that entry meant, but that clears things up.

File: src/rt_core_v2/rttuple.py (hunk 6 above)
Comment (mrw8) [reply src/rt_core_v2/rttuple.py:168]: This was whenever I flattened the inheritance hierarchy, so I removed most of the abstract tuple classes.

File: src/rt_core_v2/rttuple.py (hunk 10 above)
Comment (mrw8) [reply src/rt_core_v2/rttuple.py:442]: I'll change that, as that is a refactoring artifact that accidentally made its way in.

File: src/rt_core_v2/rttuple_formatter.py (hunk 12 above)
Comment (mrw8) [reply src/rt_core_v2/rttuple_formatter.py:9]: > json.dumps(rttuple.**__dict__**)

This will not work, as Rui is not json serializable. The options are to either have a custom function in each class to let them to decide how they will be converted to a json or (the better option) to use a custom json encoder. 

I will swap to a custom encoder to remove the dependency in a future version.
